Created Launcher subsystem

iamawesomecat opened the pull request:
added multiple constant variables for shoot speed, added 3 motors and set portID constants.

iamawesomecat committed:
Created Shooter subsystem

added multiple constant variables for shoot speed, added 3 motors and set portID constants.

iamawesomecat committed:
spotless

iamawesomecat committed:
Began methods wow

made driveMotor, StopMotor, Shoot, and initialized motors 😲

iamawesomecat committed:
implemented Shooter angle control

added 2 new methods, Get angle and set angle for the angle motor

iamawesomecat committed:
added some comments for getAngle

womp womp

kirbt commented on src/main/java/frc/team2412/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -0,0 +1,61 @@
+	public static final double ANGLE_CHANGE_SPEED = 0.15;
+
+	// HARDWARE
+	private final TalonFX shooterTopMotor;
```
I would take away the final modifier from these hardware objects. Having final here will make it so that you are unable to modify the variables and thus even though it doesn't show any errors when you attempt to, you will not be able to actually set the speed of the motors, etc.

kirbt replied:
Its good to use final on constants though, since they aren't supposed to be modified or changed

TAKBS2412 replied:
Final only means that the variable can't be assigned to another value, so operations like setting the speed of the motors will still work. ([Here's an example](https://github.com/robototes/ChargedUp2023/blob/main/src/main/java/frc/team2412/robot/subsystems/ArmSubsystem.java#L163-L165) from last year where we used final on motor controller objects).

iamawesomecat committed:
Incremented hood

added hood methods, may be removed if hood is not mechanically added. also fixed final bug in the hardware variables

iamawesomecat committed:
Renamed

Renamed everything from shooter to launcher

iamawesomecat committed:
spotless

spot less

iamawesomecat committed:
hardware finality

made hardware final (for the 2nd time)

iamawesomecat committed:
updated motors

Set all motors from talonFX to rev lol

iamawesomecat committed:
Update simgui.json

iamawesomecat committed:
spotless 😀😀😀😀😁😁😁😁😁

iamawesomecat committed:
"yOu HavE tO CaPitaLiZe pID!!11!"

iamawesomecat committed:
Started ShuffleBoard!!

Made a generic entry for launcherSpeed 😲😲😲😲😲😲

Additionally, began distanceModifier

Also fixed pid lol (kinda, still have to assign P,I,D)

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,87 @@
+	public LauncherSubsystem() {
+
+		// MOTOR INSTANCE VARIBLES
+		launcherTopMotor = new CANSparkFlex(Hardware.LAUNCHER_TOP_MOTOR_ID, MotorType.kBrushless);
+		launcherBottomMotor = new CANSparkFlex(Hardware.LAUNCHER_BOTTOM_MOTOR_ID, MotorType.kBrushless);
+		launcherAngleMotor = new CANSparkFlex(Hardware.LAUNCHER_ANGLE_MOTOR_ID, MotorType.kBrushless);
+		launcherHoodMotor = new CANSparkFlex(Hardware.LAUNCHER_HOOD_MOTOR_ID, MotorType.kBrushless);
```
Need to configure the motors. See https://github.com/robototes/ChargedUp2023/blob/f25191718390d11df9ddda789a0f0d5c35686734/src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java#L124 for an example how

iamawesomecat replied:
i did that today, thanks!

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,87 @@
+
+	public void shoot(double speed) {
+		launcherTopMotor.set(speed);
+		launcherBottomMotor.set(-speed);
```
Better to invert the bottom motor & set both to the same target speed

jamesdooley4 replied:
Even better to not set a raw power (which results in different velocities in each motor) and instead get the motor's PID controller and call setReference(value, ControlType.kVelocity) to tell the motors to run at the same velocity

iamawesomecat replied:
Ill look into doing this, as for PID for the shooter ill get into that some time this week.

jamesdooley4 replied:
Then for now, just invert the bottom motor & set it up as a follower of the top motor, so this code only needs to set the top motor power & the bottom motor will automatically follow.

jamesdooley4 replied:
And file an issue to track the follow-up work to change from power units to velocity units with PIDF

TAKBS2412 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,87 @@
+	}
+
+	public void DistanceModifier(double distance, double modifier) {
+		double speed = distance / modifier;
```
Unfinished?

iamawesomecat replied:
removed for now, ty

TAKBS2412 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,87 @@
+	private final SparkPIDController launcherAnglePIDController;
+	private final SparkPIDController launcherHoodPIDController;
+
+	GenericEntry launcherSpeed =
```
Not used - was the plan to add logging later?

iamawesomecat replied:
fixed (i think, you can add a new issue if im wrong)

iamawesomecat committed:
Config motors

you can configure the motors now

iamawesomecat committed:
spotless🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂

iamawesomecat committed:
Made first 2024 command (wow)

👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍

LauncherPreSetAngle command sets the angle of the launcher using a input when it is called

iamawesomecat committed:
spotless

guys

its without spot

iamawesomecat committed:
more shuffleboarding

iamawesomecat committed:
added note detection (wow)

thats wild

iamawesomecat committed:
spotless 😎😎😎😎😎😎😎😎😎😎😎😎😎

iamawesomecat committed:
removed distance modifier for now

:C

jamesdooley4 commented on src/main/java/frc/team2412/robot/Hardware.java:
```diff
@@ -9,8 +9,14 @@ public class Hardware {
 
+	// sensors [1 - 19]
+	public static final int LAUNCHER_SENSOR_1 = 1;
+	public static final int LAUNCHER_SENSOR_2 = 2;
```
It'd be better to name these based on where they are on the robot

iamawesomecat replied:
Yeah, we will change it once we have a idea on every sensor on the bot and number it accordingly, ty!

jamesdooley4 commented on src/main/java/frc/team2412/robot/Subsystems.java:
```diff
@@ -12,17 +13,21 @@ public static class SubsystemConstants {
 	}
 
 	public final DrivebaseSubsystem drivebaseSubsystem;
+	public LauncherSubsystem launcherSubsystem;
```
Should be `final`

iamawesomecat replied:
done, ty!

jamesdooley4 commented on src/main/java/frc/team2412/robot/commands/launcher/LauncherPreSetAngle.java:
```diff
@@ -0,0 +1,27 @@
+import edu.wpi.first.wpilibj2.command.Command;
+import frc.team2412.robot.subsystems.LauncherSubsystem;
+
+public class LauncherPreSetAngle extends Command {
```
This is named like a thing/noun instead of a action verb, which is confusing. Consider naming this `SetLauncherAngle` instead.

iamawesomecat replied:
Updated, ty!

jamesdooley4 replied:
? The class name hasn't changed...

iamawesomecat replied:
oops, fixed it now

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,140 @@
+public class LauncherSubsystem extends SubsystemBase {
+	// CONSTANTS
+	// MOTOR SPEED VALUES
+	public static final double SPEAKER_SHOOT_SPEED = 0.5;
```
These seem like power levels (% of current battery voltage, which varies over a match) and not velocity values (which are the same regardless of battery voltage). It'd be more consistent to use velocity units here.

iamawesomecat replied:
we will do this once we establish PID for the flywheels, but ty for letting me know!

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,140 @@
+		launcherHoodMotor.restoreFactoryDefaults();
+
+		// idle mode (wow)
+		launcherTopMotor.setIdleMode(IdleMode.kBrake);
```
The flywheel motors should be set to coast mode so they smoothly spin down when power is set to 0 vs. doing an abrupt stop (which uses a lot of energy to brake the flywheel).

iamawesomecat replied:
done, thanks!

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,140 @@
+
+		// current limit
+		launcherTopMotor.setSmartCurrentLimit(20);
+		launcherBottomMotor.setSmartCurrentLimit(20);
```
Likely should set the bottom motor to run inverted from the top motor, and also have it follow the top motor's settings so they run at the same velocity. Though disregard if there's a need to run the top & bottom motors at different velocities.

iamawesomecat replied:
There will be a need to run them at different velocities because of the way we want to score AMP unfortunately

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,140 @@
+
+	@Override
+	public void periodic() {
+		launcherSpeed.setDouble(launcherTopEncoder.getVelocity());
```
This is reporting in different units (velocity) than the motor's being set at (raw power). It'd be better to be consistent and use power units here for now.

iamawesomecat replied:
alright, set it to getVelocity for now, also added a comment to remind us to fix it once we establish that.

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,140 @@
+	public static final double ANGLE_TOLERANCE = 0.5;
+
+	// HARDWARE
+	private final CANSparkFlex launcherTopMotor;
+	private final CANSparkFlex launcherBottomMotor;
```
To match the current prototype & likely robot launcher, I'd recommend also adding at least one (or two) motors for the feed rollers. Consider adding a `feedTopMotor` and `feedBottomMotor` with associated methods to set their power & report it to the dashboard.

iamawesomecat replied:
talked about during meeting

jamesdooley4 replied:
FYI, in a later conversation with Marcus, there are two halves of index - one attached to the intake & one attached to the launcher. But that part of the robot was also going under a lot of design changes, so it's fine for now to leave those motors out from here.

iamawesomecat committed:
inverted bottom motor

iamawesomecat committed:
issue fixing

iamawesomecat committed:
spotless 😻🕵️‍♂️🕵️‍♂️

iamawesomecat committed:
remove sensors

:C

iamawesomecat committed:
spotless 💀💀💀💀💀💀💀💀💀💀💀💀💀💀💀💀

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,133 @@
+		launcherBottomEncoder = launcherBottomMotor.getEncoder();
+		launcherAngleEncoder = launcherAngleMotor.getAbsoluteEncoder(Type.kDutyCycle);
+		launcherHoodEncoder =
+				launcherHoodMotor.getAbsoluteEncoder(SparkAbsoluteEncoder.Type.kDutyCycle);
```
Should be able to remove `SparkAbsoluteEncoder.` here

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,133 @@
+	}
+
+	// stop specific motor method
+	public void stopMotor(CANSparkFlex motor) {
+		motor.stopMotor();
+	}
```
Shouldn't be public, as nothing outside this class should be passing a motor as a parameter.  Since it's never used, consider deleting it, or replacing with something like `stopLauncher()` that sets the two launcher motors to 0 power.

TAKBS2412 replied:
Agreed - though if we do keep it, we should make sure the motors aren't trying to fully stop (which could cause physical damage to the system) and are instead not contributing any more kinetic energy to the system (thinking back to 2022). If we do go down that route `stopMotor()` should work fine provided the other refactoring advice is followed.

iamawesomecat replied:
i went with the stopLauncher idea, ty!

TAKBS2412 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,133 @@
+	private final SparkPIDController launcherAnglePIDController;
+	private final SparkPIDController launcherHoodPIDController;
+
+	GenericEntry launcherSpeed =
```
Nit: this variable and the other `GenericEntry` should be `private final`

iamawesomecat replied:
i gotchu

iamawesomecat committed:
HOOD L

the hoods gone

iamawesomecat committed:
no more hood ID lol

iamawesomecat committed:
spotless [5:59 PM] Iyer, Darsh (Student) 🥸

darsh's emoji didnt work guys

iamawesomecat committed:
launcher angle was set

frfr

iamawesomecat committed:
Created stopLauncher (replaced stopMotor)

Set the genericEntries to private final

and replaced stopmotor with stopLauncher

iamawesomecat committed:
bunch of logging

iamawesomecat committed:
spotless 👻👻👻👻👻👻👻

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,115 @@
+
+	@Override
+	public void periodic() {
+		// .get will be replaced with .getVelocity once PID is established for flywheels :C
+		launcherSpeedEntry.setDouble(launcherTopMotor.get());
```
When the launcher stops, I think this will set the launcher setpoint used by `shoot()` to zero, which is unexpected. It'd be better to have separate dashboard entries for the current speed/power, and the desired speed/power used when `shoot()` is called.

iamawesomecat replied:
How would i get the current speed then?

iamawesomecat replied:
i think i figured this out, ty!

iamawesomecat committed:
replaced .set with .stopMotor

im convinced ill be switching this in a week

iamawesomecat committed:
added a launcher speed genericEntry

LauncherSpeed is a generic entry that displays the speed that the motors are traveling at.
LauncherSpeedEntry tracks and allows the user to change the setpoint for the speeds.

iamawesomecat committed:
command preset work

iamawesomecat committed:
the spot is less (spotless)

iamawesomecat committed:
removed thing from setAnglelaunchCommand

aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa

KangarooKoala commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,121 @@
+	}
+
+	public void launch() {
+		launcherTopMotor.set(launcherSpeedEntry.getDouble(SPEAKER_SHOOT_SPEED));
+		launcherBottomMotor.set(launcherSpeedEntry.getDouble(SPEAKER_SHOOT_SPEED));
```
```suggestion
        double speed = launcherSpeedEntry.getDouble(SPEAKER_SHOOT_SPEED);
		launcherTopMotor.set(speed);
		launcherBottomMotor.set(speed);
```
This guarantees that the the motors are set to the same speed (in case the entry gets set to a different value between the calls).

iamawesomecat replied:
Okay, i did exactly that, tysm!

KangarooKoala commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,121 @@
+					.withSize(1, 1)
+					.withWidget(BuiltInWidgets.kTextView)
+					.getEntry();
+	private final GenericEntry launcherSpeed =
```
Please pick more descriptive names for this and `launcherSpeedEntry` to clarify that one is set by the user and the other is set by robot code.

iamawesomecat replied:
Fixed, ty

KangarooKoala commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,121 @@
+
+	private final GenericEntry launcherAngleEntry =
+			Shuffleboard.getTab("Launcher")
+					.addPersistent("Launcher angle", getAngle())
```
Should this and `launcherSpeed` be persistent? I don't think we want the value to be saved across runs of the robot, and I could imagine this causing confusion by hiding when the robot isn't connected.

iamawesomecat replied:
no it shouldnt lol, thats my bad. fixed now tho :D

KangarooKoala commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -0,0 +1,121 @@
+		launcherAngleMotor.burnFlash();
+	}
+
+	// stop specific motor method
+	public void stopLauncher() {
```
Please either update the comment or delete it.

iamawesomecat replied:
done, ty

KangarooKoala commented on src/main/java/frc/team2412/robot/commands/launcher/LaunchCommand.java:
As a note, because this command only sets the motor speeds when it (the command) is initialized, changes to the launcher speed setpoint entry will not be updated until the command is scheduled again.

iamawesomecat replied:
fixed, ty!

TAKBS2412 reviewed (dismissed):
LGTM pending Joseph's comments (note the PR will have to be re-approved as adding new commits removes existing approvals).

iamawesomecat committed:
fixed some issues lol + spotless

iamawesomecat committed:
capitalization 💯💯💯💯💯💯💯💯💯💯

iamawesomecat committed:
Change file capitalization

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -33,7 +33,7 @@ public class LauncherSubsystem extends SubsystemBase {
 	private final SparkPIDController launcherAnglePIDController;
 
-	private final GenericEntry launcherSpeedEntry =
+	private final GenericEntry SetlauncherSpeedEntry =
```
Probably should be lowercase to start, to match the other class variables: `setLauncherSpeedEntry`

iamawesomecat committed:
made another launch method for the heck of it lol

iamawesomecat committed:
capitalization moment

TAKBS2412 commented on src/main/java/frc/team2412/robot/subsystems/LauncherSubsystem.java:
```diff
@@ -103,6 +103,11 @@ public void launch() {
 		launcherTopMotor.set(speed);
 		launcherBottomMotor.set(speed);
```
(this could be renamed and rewritten to use the other `launch()` method... but I think these methods will change anyway once we switch to velocity instead of power, so probably fine to leave for now).

iamawesomecat committed:
spotless 🗽🗽🗽🗽🗽🗽
